### ISSUE TITLE
Keep track of the AST nodes associated with each code path segment

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#10654]
**The version of ESLint you are using.**

5.2.0

**The problem you want to solve.**

When using the code path analysis API, it's very cumbersome to perform graph traversal operations on a control flow graph because `CodePathSegments` objects don't retain information about which AST nodes they refer to.

As an example, I'm currently implementing https://github.com/eslint/eslint/issues/10405. I'd like to report an error under roughly the following circumstances:

> There is some acyclic path on the control flow graph where all of the following steps happen in order, possibly with other steps between them:
> 1. Control flow enters an `AssignmentExpression` that assigns to a variable.
> 1. Control flow enters an `Identifier` which maps to the same variable that is assigned in the `AssignmentExpression`.
> 1. Control flow exits an `AwaitExpression` or `YieldExpression`.
> 1. Control flow exits the `AssignmentExpression` from (1).

Intuitively, this idea seems like it should be relatively simple to implement. Ideally, I'd be able to create a function that accepts a `CodePath` and performs a graph traversal on its `CodePathSegment`s, while iterating through the AST nodes in each `CodePathSegment` to find the specific nodes that I'm searching for.

Unfortunately, the implementation is significantly more complicated with the existing code path API. A rule can only get information about code path segments when they start or end (i.e. at branch points). In order to determine which code path segment is enabled for a given node, a rule needs to manually keep track of the "current" code path segment and refer to it in AST listeners.

**Your take on the correct solution to problem.**

I think each `CodePathSegment` should have an array property containing the nodes that it enters and exits in the path. For example, it might look something like this:

```js
nodes: [
    {
        entering: true,
        node: <First node>
    },
    {
        entering: false,
        node: <First node>
    },
    {
        entering: true,
        node: <Second node>
    },
    {
        entering: true,
        node: <Third node>
    },
    {
        entering: false,
        node: <Third node>
    },
    {
        entering: false,
        node: <Second node>
    }
]
```

In other words, I want rules to have access to the nodes that are displayed in the control flow graphs on [this page](https://eslint.org/docs/developer-guide/code-path-analysis). In my opinion, this would make the code path analysis API significantly easier to use.

cc @mysticatea
